test/unit/master/test_build.py

```diff
@@ -273,6 +273,26 @@ def test_cancel_is_a_noop_if_build_is_already_finished(self):
         self.assertEqual(len(mock_slave.method_calls), num_slave_calls_before_cancel,
                          'Canceling a finished build should not cause any further calls to slave.')
 
+    def test_prepare_callback_handles_atomizererror_exception_and_raises_again(self):
+        job_config = self._create_job_config()
+        subjobs = self._create_subjobs(job_config=job_config)
+        mock_subjob_calculator = self._create_mock_subjob_calc(subjobs, compute_subjob_raises_exception=True)
+        build = self._create_test_build(BuildStatus.QUEUED)
+
+        with self.assertRaises(AtomizerError):
+            build.prepare(mock_subjob_calculator)
+
+    def test_cancel_before_prepare_does_not_call_compute_subjobs(self):
+        job_config = self._create_job_config()
+        subjobs = self._create_subjobs(job_config=job_config)
+        mock_subjob_calculator = self._create_mock_subjob_calc(subjobs)
+        build = self._create_test_build(BuildStatus.QUEUED)
+
+        build.cancel()
```

Comment:
But I don't think this tests anything otherwise. As this test is currently written, `mock_subjob_calculator` has nothing to do with the build object (since it is never passed to prepare()).

It would be impossible for this test to ever fail since `mock_subjob_calculator` is never passed to the build. In my opinion the only valuable thing to test is that `compute_subjobs_for_build` is not called if `build.cancel()` happens before `build.prepare()`.